### PR TITLE
Closing

### LINKED ISSUE #2318: Morphologically close a volume

### PR DESCRIPTION
**Pull request by 0x00b1**
Resolves #2318.

### CODE REVIEW

File: cellprofiler/modules/closing.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+"""
+
+import cellprofiler.image
+import cellprofiler.module
+import cellprofiler.setting
+import skimage.morphology
+
+
+class Erosion(cellprofiler.module.Module):
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: Rename `Closing`.

File: cellprofiler/modules/closing.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    variable_revision_number = 1
+
+    def create_settings(self):
+        self.x_name = cellprofiler.setting.ImageNameSubscriber(
+            "Input"
+        )
+
+        self.y_name = cellprofiler.setting.ImageNameProvider(
+            "Output",
+            "OutputImage"
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: Maybe `"Closing"` or `"Closed"`

File: cellprofiler/modules/closing.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+"""
+
+import cellprofiler.image
+import cellprofiler.module
+import cellprofiler.setting
+import skimage.morphology
+
+
+class Erosion(cellprofiler.module.Module):
+    category = "Mathematical morphology"
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: Maybe `"Morphology"`?

File: cellprofiler/modules/closing.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+        x = images.get_image(x_name)
+
+        x_data = x.pixel_data
+
+        y_data = skimage.morphology.closing(x_data, structuring_element)
+
+        y = cellprofiler.image.Image(
+            image=y_data,
+            parent_image=x
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: Output image needs `dimensionality`.

File: cellprofiler/modules/closing.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+        images.add(y_name, y)
+
+        if self.show_window:
+            workspace.display_data.x_data = x_data
+
+            workspace.display_data.y_data = y_data
+
+    def display(self, workspace, figure):
+        figure.set_grids((1, 2))
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: Support 2D display.

**Comment by mcquin** [comment]
Needs documentation and tests.

File: cellprofiler/modules/closing.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+        if self.show_window:
+            workspace.display_data.x_data = x_data
+
+            workspace.display_data.y_data = y_data
+            
+            workspace.display_data.dimensions = dimensions
+
+    def display(self, workspace, figure):
+        figure.set_subplots((1, 2))
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: Needs `dimensions`.

File: cellprofiler/modules/closing.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
+"""
+
+Closing
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: Documentation

File: tests/modules/test_closing.py [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import cellprofiler.modules.closing
+import cellprofiler.workspace
+import numpy.testing
+import skimage.data
+import skimage.morphology
+
+instance = cellprofiler.modules.closing.Closing()
```
Comment (mcquin) [inline tests/modules/test_closing.py]: So cool!

File: tests/modules/test_closing.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
+import cellprofiler.modules.closing
+import cellprofiler.workspace
```
Comment (mcquin) [inline tests/modules/test_closing.py]: Unused

File: cellprofiler/modules/closing.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
+# -*- coding: utf-8 -*-
+
+"""
+
+Closing
+
+The erosion of the dilation of an image. It’s used to remove pepper noise.
+
+A structuring element is a shape that’s used to interact with an image.
```
Comment (mcquin) [inline cellprofiler/modules/closing.py]: I think this should go on the setting's help.

File: cellprofiler/modules/closing.py (hunk 6 above)
Comment (mcquin) [inline cellprofiler/modules/closing.py]: `dimensions` are needed here

File: tests/modules/test_closing.py (hunk 9 above)
Comment (mcquin) [inline tests/modules/test_closing.py]: ^ bump

File: cellprofiler/modules/closing.py (hunk 10 above)
Comment (0x00b1) [inline cellprofiler/modules/closing.py]: I think this should be a default for that setting. 🤓

**Pull request merged**